pkg/agent/agent.go

```diff
@@ -0,0 +1,450 @@
+	klog.Infof("node: %+v", a.Options.NodeName)
+
+	klog.V(2).Info("setting up event broadcaster")
+
+	recorder := a.broadcaster.NewRecorder(scheme.Scheme, metadata.EventsReporter)
+
+	a.broadcaster.StartRecordingToSink(ctx.Done())
+	defer a.broadcaster.Shutdown()
+
+	klog.V(2).Info("setting up periodic reconciliation ticker")
+
+	ticker := time.NewTicker(a.ReconcileInterval)
+	defer ticker.Stop()
+
+	drbdUpdateResult := make(chan error)
+	a.drbdResources = NewDrbdResources(a.DrbdStatusInterval)
+	go func() {
+		defer close(drbdUpdateResult)
+		drbdUpdateResult <- a.drbdResources.StartUpdates(ctx)
+	}()
+
+	a.informerFactory.Start(ctx.Done())
+
+	for {
+		err := a.reconcile(ctx, recorder)
+		if err != nil {
+			klog.V(1).ErrorS(err, "error during reconciliation")
+		}
+
+		select {
+		case <-ctx.Done():
+			klog.V(2).Info("context done")
+			return ctx.Err()
+		case err := <-drbdUpdateResult:
+			klog.V(2).Info("drbd syncer done")
+			return err
+		case <-ticker.C:
+			klog.V(2).Info("starting periodic reconciliation")
+		}
+	}
+}
+
+// reconcile the expected cluster state with the known cluster state.
+func (a *agent) reconcile(ctx context.Context, recorder events.EventRecorder) error {
+	klog.V(1).Info("starting reconciliation")
+
+	now := time.Now()
+	resourceState := a.drbdResources.Get()
+
+	var wg sync.WaitGroup
+
+	// Manage own node taints
+
+	wg.Add(1)
+	taintsCtx, cancel := context.WithTimeout(ctx, a.Timeout())
+	defer cancel()
+
+	go func() {
+		defer wg.Done()
+
+		err := a.ManageOwnTaints(taintsCtx, resourceState, now, recorder)
+		if err != nil {
+			klog.V(1).Infof("managing node taints failed: %s", err)
+		}
+
+		klog.V(2).Infof("Own node taints synced")
+	}()
+
+	for i := range resourceState {
+		resource := &resourceState[i]
+
+		klog.V(2).Infof("checking on resource %s", resource.Name)
+
+		req, err := a.getRequestForDrbdResource(resource, now)
+		if err != nil {
+			klog.V(2).Infof("ignoring resource: %w", err)
+			continue
+		}
+
+		for i := range a.reconcilers {
+			r := a.reconcilers[i]
+			wg.Add(1)
+
+			go func() {
+				defer wg.Done()
+
+				ctx, cancel := context.WithTimeout(ctx, a.Timeout())
+				defer cancel()
+
+				err := r.RunForResource(ctx, req, recorder)
+				if err != nil {
+					klog.V(1).Infof("reconciliation failed: %v", err)
+				}
+			}()
+		}
+	}
+
+	wg.Wait()
+
+	return nil
+}
+
+// ManageOwnTaints ensures that the taints on the local node are up to date:
+// * If all resources are in a good state, remove all taints.
+// * Add taints for DRBD resources that are forcing IO errors.
+func (a *agent) ManageOwnTaints(ctx context.Context, resourceState []DrbdResourceState, refTime time.Time, recorder events.EventRecorder) error {
+	allQuorum := true
+	anyForceIOError := false
+	for i := range resourceState {
+		if resourceState[i].ForceIoFailures {
+			anyForceIOError = true
+		}
+
+		for j := range resourceState[i].Devices {
+			if !resourceState[i].Devices[j].Quorum {
+				allQuorum = false
+			}
+		}
```

Comment:
Since we check for 2 conditions in the same loop, it could get messy. I don't think the extra checks are worth it. In most cases we have to run through the whole loop anyways (in case that everything is fine).